src/trace.rs

```diff
@@ -2,7 +2,11 @@
 /// [`Trace::trace`](trait.Trace.html#method.trace).
 pub type Tracer<'a> = dyn FnMut(*const ()) + 'a;
 
-/// Defines how the cycle collector should collect a type.
+/// Defines how the cycle collector should collect a type. Implementors
+/// must ensure that the [`Drop`] impl of any type that implements this trait
+/// does not modify or allow any garbage collected pointers (['Cc<T>']) to
```

Comment:
I found this a bit vague to understand - unclear what "modify" or "escape" means exactly and what's the consequences if they do. It also seems to suggest that one might have to impl `Drop` manually while I think most of cases the default `Drop` impl is good enough.

The docstring of the `trace` function below has text:

    /// The values that are visited must match the `Drop::drop`
    /// implementation. Otherwise memory leak or panic might
    /// implementation. Otherwise memory leak or panic might
    /// happen. After the panic, dereferencing already collected
    /// happen.

Is it basically the same thing? If so, I'd like to keep the trait doc-string high-level and update `trace` if necessary.

Reply:
I'm basically trying to say this this isn't a finalizer - if you resurrect objects you'll get a panic when we go over and sanity check the refcounts.

Reply:
Got it. It seems there are 2 forbidden patterns:
- Accessing the value (dereferncing `Cc<T>`). This is forbidden because `T` might be dropped already.
- Conditionally cloning (resurrecting)  `Cc<T>`. This seems okay if the `trace()` function skips visiting that `Cc<T>` with the same condiiton.

Reply:
How about something like:

```
/// ## Customized `Drop` implementation
///
/// The [`Drop`] implementation should avoid dereferencing other [`Cc<T>`]
/// objects.  Failing to do so might cause panic or undefined behavior. For
/// example, `T1` has a field `Cc<T2>`. The collector might have already
/// dropped `T2` by the time `T1::drop` runs.
///
/// The `Drop` implementation should also be careful about cloning
/// (resurrecting) [`Cc<T>`] objects. If it must do so, the `trace`
/// implementation should match by avoiding visiting those cloned objects.
pub trait Trace: 'static {
...
```

?

Reply:
I like that wording a lot - pushed!